Add EIP: Hash-Chain RANDAO (#11861)
ethereum/EIPs: EIPS/eip-8321.md

---
eip: 8321
title: Hash-Chain RANDAO
description: Replace the BLS-signature RANDAO reveal with a post-quantum hash-chain commit-reveal scheme
author: Kevaundray Wedderburn (@kevaundray), Benedikt Wagner (@benedikt-wagner), Tom Wambsgans (@TomWambsgans), Justin Drake (@JustinDrake)

> discussions-to: https://ethereum-magicians.org/t/hash-chain-randao/28942

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-8321-hash-chain-randao/28942
```
Updated Eth Magicians title with assigned number

status: Draft
type: Standards Track
category: Core
created: 2026-07-05
---

## Abstract

> Replace the BLS-signature-based RANDAO reveal with a hash-chain commit-reveal scheme. RANDAO's resistance to grinding currently relies on BLS signatures being *unique*, so that a proposer cannot bias its contribution. Since BLS relies on pre-quantum hardness assumptions, a quantum computer can recover a validator's key and predict its future reveals.

Reviewer 2: ```suggestion
Replace the BLS-signature-based Random Decentralized Autonomous Organization (RANDAO) reveal with a hash-chain commit-reveal scheme. RANDAO's resistance to grinding currently relies on BLS signatures being *unique*, so that a proposer cannot bias its contribution. Since BLS relies on pre-quantum hardness assumptions, a quantum computer can recover a validator's key and predict its future reveals.
```

A commitment is registered once, via a new per-block-capped beacon operation (similar to `BLSToExecutionChange`); it cannot be updated in place, so a validator that ever needs a new chain exits and re-enters. Validators that have not yet registered a commitment continue to use the legacy BLS reveal; this is a transitional path intended for removal in a later fork.

We note the simplicity: the protocol holds only the current commitment (32 bytes) and walks one link back per proposal for that validator.

## Motivation

The primary driver is post-quantum readiness.

Today's RANDAO contribution is a BLS signature over the epoch number. Its security rests on the *uniqueness* of BLS signatures: given the message and public key, exactly one valid signature exists, so a proposer cannot grind their own contribution to bias the randomness.

> BLS is not post-quantum safe, so a cryptographically relevant quantum computer (CRQC) can recover a validator's secret key from its public key. Uniqueness still holds, but the attacker can compute the validator's reveals itself, predicting the chain's future randomness, and thus the proposer schedule, far in advance.

Reviewer 3: I might be worth mentioning that not only can the attacker predict the schedule in advance, but it can also influence it (much better than without a CRQC): The standard reveal-or-not attack usually requires the attacker to have $k$ consecutive proposer slots at the end of an epoch to choose from $2^k$ options. If the attacker can predict others' randomness contributions, they can now user any $k$ slots (assuming all honest parties always reveal) ; they do not need to be consecutive or at the end of an epoch.

> Most uses of signatures, such as block proposals and attestations, can move to any secure post-quantum scheme. RANDAO, however, additionally depends on signature *uniqueness*, and most post-quantum schemes do not provide it, allowing one to grind for a favorable randao contribution. The lean signature scheme currently planned for the consensus layer is grindable in exactly this way, since it includes a *salt* component.

Reviewer 4: Here you mention lean signature, this more the name that we give internally, shouldn't we precise XMSS since it's more the real name or maybe this is intentional to keep things generic?

> Most uses of signatures, such as block proposals and attestations, can move to any secure post-quantum scheme. RANDAO, however, additionally depends on signature *uniqueness*, and most post-quantum schemes do not provide it, allowing one to grind for a favorable randao contribution. The lean signature scheme currently planned for the consensus layer is grindable in exactly this way, since it includes a *salt* component.

Reviewer 2: ```suggestion
Most uses of signatures, such as block proposals and attestations, can move to any secure post-quantum scheme. RANDAO, however, additionally depends on signature *uniqueness*, and most post-quantum schemes do not provide it, allowing one to grind for a favorable RANDAO contribution. The lean signature scheme currently planned for the consensus layer is grindable in exactly this way, since it includes a *salt* component.
```

The commit-reveal structure also preserves RANDAO's existing security model, since the whole chain is fixed at commitment time, long before the validator knows its proposal slots or the mixes it might want to bias, so the proposer's only remaining lever is the same as the one it has today: reveal, or withhold and forfeit the block.

Note: This EIP does not make beacon chain randomness post-quantum secure end to end; block signatures and the registration operation's signature still use BLS. It incrementally introduces a post-quantum version of RANDAO, so that a later post-quantum fork which changes the signature-scheme and sets the initial hash-chain commitment at deposit time can complete the transition.

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Reviewer 2: RFC 2119 and RFC 8174 should be linked (EIP-1 IETF-link convention), and this boilerplate is also missing `NOT RECOMMENDED`. Suggested standard form:

```suggestion
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
```

(at least link the RFCs)

> These changes are applied to the consensus specifications at a fork to be scheduled.

Reviewer 2: The Specification uses functions and constants inherited from the consensus specs (`get_randao_mix`, `compute_signing_root`, `get_domain`, `MIN_SEED_LOOKAHEAD`, `G2_POINT_AT_INFINITY`, `BLSToExecutionChange`, `set_or_append_list`, ...) without stating where they are defined. Following the EIP-3675 precedent, please name the repository once, e.g. "These changes are applied to the consensus specifications (`ethereum/consensus-specs`) at a fork to be scheduled; functions and constants not defined here retain their meaning from those specifications."

### Hash chain construction (off-chain)

A validator generates a uniformly random chain secret `c_0` and computes a chain of length `n`:

```text

> c_i = hash(HASH_CHAIN_RANDAO_DST + c_{i-1})    for i in 1..n

Reviewer 2: `HASH_CHAIN_RANDAO_DST` is used before definition, which is somewhat confusing. Either move the Constants section above this text, or define the type of this constant so it is more clear what is intended. The reader will now likely jump forward to figure out what this constant is, which is not desired.

> | `HASH_CHAIN_RANDAO_DST` | `b'HASH_CHAIN_RANDAO'` | Domain-separation tag prefixed to each hash-chain link |

Reviewer 3: I would suggest including the EIP-155 ChainID in the domain separation string.

```

where `hash` is SHA-256, as elsewhere in the consensus specifications. `HASH_CHAIN_RANDAO_DST` is a fixed domain-separation prefix, so hash-chain links cannot collide with hashes used elsewhere in the protocol. It carries no per-link index, so a validator still needs to store only chain values, not their positions.

> The chain secret `c_0` MUST NOT be the zero word: a zero reveal is rejected by `process_randao` since that value is used to indicate a validator that has not registered a hash chain.

Reviewer 5: Maybe we should even say that the validator MUST check that no `c_i` is zero, and generate a new chain in case this happens. Of course, it happens with negl probability (so practically never), but it is worth explaining this somewhere.

| `COMMITMENT_REGISTRATION_DELAY` | `Epoch(3)` | Epochs before a new commitment becomes active; MUST be at least `MIN_SEED_LOOKAHEAD + 2` (see Rationale) |
| `MAX_RANDAO_COMMITMENT_REGISTRATIONS` | `uint64(128)` | Maximum commitment registration operations per block |
| `PENDING_RANDAO_COMMITMENTS_LIMIT` | `uint64(2**20)` | Bound on the pending commitment queue. This is in practice not reachable and is a SSZ type bound only |

> ### Hash chain construction (off-chain)

Reviewer 2: Nit: CONTRIBUTING.md asks for title case in headings — several are sentence case ("Hash chain construction (off-chain)", "Chain exhaustion", "Block processing", "Epoch processing", "Fork transition", "Sunset of the legacy path", and the Rationale / Security Considerations subheadings).

> A validator generates a uniformly random chain secret `c_0` and computes a chain of length `n`:

Reviewer 3: This does not specify the data type of `c_0` (in particular, the bitlength). In principle, `c_0` could be longer/shorter than the other `c_i`'s (but then you can't reveal it and there is no good reason to do so). So I guess for clarity, we should specify it.

```text
c_i = hash(HASH_CHAIN_RANDAO_DST + c_{i-1})    for i in 1..n
```

> where `hash` is SHA-256, as elsewhere in the consensus specifications. `HASH_CHAIN_RANDAO_DST` is the fixed byte-string domain-separation prefix defined in the Constants section above, so hash-chain links cannot collide with hashes used elsewhere in the protocol. It carries no per-link index, so a validator still needs to store only chain values, not their positions.

Reviewer 4: Here you mention SHA-256 but potentially this could change to BLAKE3 as per the Flock/binary field results right? Maybe we should say that somehow?

Author: Chnged to blake3 -- the CL uses sha a lot :D

> No chain value may be the zero word: `process_randao` rejects a zero reveal (the zero word marks an unregistered validator), and a zero commitment cannot be registered. The seed `c_0` is chosen non-zero, and every other `c_i` is a SHA-256 output, so a zero link occurs only with negligible probability (about `n / 2**256`). Even so, a validator MUST check that no `c_i` is the zero word and regenerate the chain if any link is zero.

Reviewer 4: "MAY" should be capital letter no?

```suggestion
No chain value MAY be the zero word: `process_randao` rejects a zero reveal (the zero word marks an unregistered validator), and a zero commitment cannot be registered. The seed `c_0` is chosen non-zero, and every other `c_i` is a SHA-256 output, so a zero link occurs only with negligible probability (about `n / 2**256`). Even so, a validator MUST check that no `c_i` is the zero word and regenerate the chain if any link is zero.
```

> With the recommended `n` this never happens in practice. A validator proposes on the order of 100 times per year even in aggressive futures, so a chain of `2**13` links lasts for centuries; sizing `n` generously (it is cheap) makes exhaustion a non-issue. The same applies to a lost or mis-generated chain secret: there is no in-place recovery, so the remedy is exit and re-entry, and the chain secret SHOULD be guarded like the signing key (see Security Considerations).

Reviewer 5: Wait: `2**13 links / (100 links / year) < 82 years` .

That's not centuries, right?

Author: Ah yes, I was going back and forth between 2^13 and 2^16

Validators SHOULD choose `n` large enough that the chain outlasts the validator, since it cannot be extended in place. `n >= 2**16` (~65,000 links) is RECOMMENDED. The whole chain can be generated in milliseconds, and storing every link takes ~2 MB. This exact value is of course arbitrary: the protocol never learns or enforces `n`, and since generation and storage stay cheap, most operators lose nothing by choosing a larger `n`.

This bounded lifetime is the one structural difference from BLS, where a public key is a commitment that never expires, a hash chain instead lasts for its length. It is not a practical constraint though since we can generate a large enough chain that will last for centuries.

#### Chain exhaustion

A commitment cannot be updated in place, so a chain cannot be extended once registered. The chain is exhausted only when its stored commitment reaches the secret seed `c_0`, whose preimage the validator does not hold. A validator whose chain runs out can no longer propose on the hash-chain path and must exit and re-enter as a new validator to obtain a fresh chain.

> With the recommended `n` this never happens in practice. A validator proposes on the order of 100 times per year even in aggressive futures, so a chain of `2**16` links lasts for centuries; sizing `n` generously (it is cheap) makes exhaustion a non-issue. The same applies to a lost or mis-generated chain secret: there is no in-place recovery, so the remedy is exit and re-entry, and the chain secret SHOULD be guarded like the signing key (see Security Considerations).

Reviewer 4: I imagine that the recommendation will be made to the client teams to have a large enough `n`. But since the fact of having to exit the validator set and re-enter it is very painful, can't we have a sort of delayed commitment rotation?

I think this adds a bit of complexity but basically what we could do is to allow a validator to register a replacement hash chain before the current one is exhausted.

For example when the validator has, say, 1,000 links left, it generates another chain and submits a fresh `RandaoCommitmentRotation` signed by its validator signing key. The protocol does not activate `d_m` immediately. It places it in a pending queue and activates it after the same safe lookahead delay used for initial registration:

```
activation_epoch = inclusion_epoch + COMMITMENT_REGISTRATION_DELAY
```

Author: Ah for this we would also need to commit the hash chain length for each validator too I think

`BeaconBlockBody` gains two fields and retains `randao_reveal` transitionally:

```python
class BeaconBlockBody(Container):
    randao_reveal: BLSSignature  # transitional; MUST be the G2 point at infinity once the proposer has an active commitment
    # ... existing fields ...
    hash_chain_reveal: Bytes32  # [New in this EIP] zero unless the proposer has an active commitment
    randao_commitment_registrations: List[SignedRandaoCommitmentRegistration, MAX_RANDAO_COMMITMENT_REGISTRATIONS]  # [New in this EIP]
```

The new fields are appended after all existing fields, following the convention since Capella of adding fields at the end.

`BeaconState` gains a commitment registry and a pending queue. `randao_commitments` is indexed by validator index, holding one entry per registry member; a zero entry means no commitment is registered and the legacy BLS reveal path applies. The `Validator` container is unchanged.

```python
class BeaconState(Container):
    # ... existing fields ...
    randao_commitments: List[Bytes32, VALIDATOR_REGISTRY_LIMIT]  # [New in this EIP]
    pending_randao_commitments: List[PendingRandaoCommitment, PENDING_RANDAO_COMMITMENTS_LIMIT]  # [New in this EIP]

> ```

Reviewer 6: don't we need to make those progressive? cc @etan-status

Reviewer 7: If PENDING_RANDAO_COMMITMENTS_LIMIT may evolve over time, or if the list is typically only partially filled with few elements, then ProgressiveList[T] would be preferrable.

#### New `queue_randao_commitment`

At most one registration per validator may be pending at any time; a block containing a registration for a validator with an in-flight one is invalid.

```python
def queue_randao_commitment(state: BeaconState, index: ValidatorIndex, commitment: Bytes32) -> None:
    # Rejects a second registration for the same validator, both across blocks and
    # within a single block: the first op appends this entry, so a later op for the
    # same index fails this assert (mirrors how voluntary exits self-reject in-block).
    assert all(pending.validator_index != index for pending in state.pending_randao_commitments)
    state.pending_randao_commitments.append(PendingRandaoCommitment(
        validator_index=index,
        commitment=commitment,
        activation_epoch=get_current_epoch(state) + COMMITMENT_REGISTRATION_DELAY,
    ))
```

The resulting lifecycle, worked through to be more explicit:

> - A registration included in epoch `N` is queued with activation epoch `N + 3`.

Reviewer 2: ```suggestion
- A registration included in epoch `N` is queued with activation epoch `N + COMMITMENT_REGISTRATION_DELAY`.
```

- The validator stays on the legacy BLS path throughout epochs `N` to `N + 2`; its proposals in that window are BLS reveals as before.

> - At the epoch transition into `N + 3`, the pending entry is consumed and the hash chain becomes active; from then on the validator reveals from its chain.

Reviewer 4: Maybe worth adding here that validators should continue using the legacy BLS path until the commitment is active in the canonical state, even if they observed their registration in a block that was later reorged out?

### Why one link per block proposed

The legacy reveal is a signature over the epoch, so two proposals by the same validator in one epoch reveal the same value. Consuming one link per block is simpler (no per-epoch bookkeeping), makes every block contribute fresh entropy, and costs nothing, since chains are cheap to generate at any reasonable length.

We also note that with consolidations, the probability of a validator proposing multiple times in the same epoch increases, effectively cancelling out their RANDAO contribution.

### Why the protocol does not track chain length

Storing only the current commitment and requiring a preimage per proposal makes the chain length a purely private, off-chain parameter. The protocol state cost is one 32-byte entry per validator regardless of chain length, and validators can size chains so that the chain outlasts the validator. The only reason to store the chain length would be to mandate a specific length for every validator, which seems unnecessarily restrictive.

### Why an activation delay on registration, and why `MIN_SEED_LOOKAHEAD + 2`

> The scheme's grinding resistance comes entirely from the commitment predating the validator's knowledge of what it would want to bias. If a validator could register a chain just before proposing and it became active instantly, it could grind: generate many candidate chains, compute the mix each would produce at the upcoming slot, and register the most favourable one. We add a delay, sized so that the registrant cannot even know *whether* it will propose in the activation epoch at the time it chooses the chain.

Reviewer 4: Here I feel that the intuition makes sense but the rationale probably doesn't prove exactly why `MIN_SEED_LOOKAHEAD + 2` is sufficient.

Maybe we could add an explicit epoch timeline showing:
- registration inclusion epoch
- when proposer duties become knowable
- activation epoch
- first slot where the new commitment maybe be used

Just because I feel that the +2 isn't obvious from the current text.

Replay-safety falls out of the unregistered check. A registration is valid only while `randao_commitments[validator_index]` is zero. Once it activates, the entry is non-zero, so any later replay of the message is invalid; and while the registration is merely pending, the one-pending rule blocks a second submission. This is the analogue of the idempotence predicate that makes `BLSToExecutionChange` replay-safe, whose handler asserts the credentials are still BLS-prefixed, which the first application falsifies.

### Why keep a legacy fallback instead of a registration deadline

There is no urgent need for validators to be fully post-quantum yet, so a hard deadline buys little; unregistered validators keep functioning on the BLS path in the meantime. There will also likely be a long tail of validators who never submit the message because they run custom software. We therefore defer the forced cutover to a fork that must touch these code paths anyway, the post-quantum switchover, which also lets us exercise the new code paths first.

## Backwards Compatibility

> This EIP requires a scheduled consensus-layer hard fork. Within the fork, the change is backwards compatible from the validator's perspective: unregistered validators continue proposing exactly as today. Downstream consumers of `randao_mixes` (including the execution layer's `PREVRANDAO`) are unaffected; the mix remains a 32-byte accumulator updated once per block, only the provenance of contributions changes.

Reviewer 2: ```suggestion
This EIP requires a scheduled consensus-layer hard fork. Within the fork, the change is backwards compatible from the validator's perspective: unregistered validators continue proposing exactly as today. Downstream consumers of `randao_mixes` (including the execution layer's [EIP-4399's](./eip-4399.md) `PREVRANDAO`) are unaffected; the mix remains a 32-byte accumulator updated once per block, only the provenance of contributions changes.
```

## Test Cases

State-transition test vectors to be provided in the consensus-specs test suite.

## Security Considerations

### Biasability is unchanged

The proposer's only degree of freedom is still withholding: reveal and propose, or withhold and forfeit the block plus its rewards. This is the same one-bit-per-proposer bias RANDAO has today, with the same economic cost. The validator freely choosing its own chain values (whereas a BLS reveal is uniquely determined by the key and epoch, leaving no freedom) does not add grinding power, because the entire chain is fixed before the validator knows its proposal slots or the co-contributions to any future mix.

### Registration cannot be used reactively

> A validator that sees a proposal duty approaching cannot register a favourable chain to influence the outcome. Registration activates `COMMITMENT_REGISTRATION_DELAY` epochs after inclusion, strictly after every slot for which duties were knowable at signing time, so the chain is always fixed before the seed that draws any duty it could serve.

Reviewer 4: Maybe here the notion of "knowable at signing time" is a bit confusing since the protocol never observes when the registration was signed no?

From my understanding, the security argument is really about the inclusion epoch, since activation is derived from inclusion rather than signing.

Perhaps we should rephrase the argument in terms of inclusion to avoid relying on an unobservable notion what do you think?

### Degenerate chains

A validator may choose pathological values for its *own* chain, for example deriving the chain secret from a publicly known constant like the genesis hash, so that all its reveals are predictable in advance and contribute no effective entropy.

This cannot be exhaustively restricted, and we do not try to, because RANDAO's security does not assume that *every* contribution is honest. If this were the case, then a validator choosing to add nothing by withholding would also break the security; one honest contributor per mixing period suffices.

We do forbid the pathological case of using the zero word, since this would collide with the unregistered sentinel when revealed, so `process_randao` rejects zero reveals.

### Chain loss and theft

Losing the chain secret makes a validator unable to produce valid blocks on the hash-chain path, and there is no in-place recovery: a commitment cannot be re-registered. The remedy is to exit and re-enter as a new validator. Funds are unaffected (withdrawal does not depend on the chain), and the failure mode is missed proposals, not slashing risk.

> Theft of the chain secret alone lets an attacker predict (not choose) the validator's future contributions, strictly less power than theft of the signing key today. Validator clients SHOULD treat the chain secret with the same custody standards as signing keys. As a safety measure, it should not be derived in a way that links it to the signing key: a hash chain, taken to its limit, eventually reveals its seed, whereas a secret key must never be revealed.

Reviewer 5: as an analogy, maybe one can say here that this is in some sense equivalent to leaking a long list of future signatures for BLS-RANDAO.

Author: this is a good intuition/analogy -- added it in

> Theft of the chain secret alone lets an attacker predict (not choose) the validator's future contributions. In effect this is equivalent to leaking the validator's entire list of future BLS-RANDAO reveals: the attacker learns every future contribution in advance, but can neither choose them nor sign anything else on the validator's behalf. It is therefore strictly less power than theft of the signing key today, which gives the attacker that same prediction plus the ability to sign the validator's blocks and attestations. Validator clients SHOULD treat the chain secret with the same custody standards as signing keys. As a safety measure, it should not be derived in a way that links it to the signing key: a hash chain, taken to its limit, eventually reveals its seed, whereas a secret key must never be revealed.

Reviewer 2: RFC 2119 keywords are only allowed in the Specification section (see CONTRIBUTING.md). This `SHOULD` is in Security Considerations, please lowercase to `should` (or move the requirement into Specification).
The reason why this is discouraged/not allowed is because this could "hide" specifications in other sections than the Specifications section.

The legacy BLS reveal signs the epoch number, so a reveal exposed in a block that never makes the canonical chain is worthless outside that epoch. A hash-chain reveal has no such time binding: a reveal published in an *orphaned* block remains the validator's mandatory next contribution at whatever future slot it next proposes. An observer who collects reveals from orphaned blocks (or losing forks) accumulates known future co-contributions.

Because a chain cannot be retired in place, there is no cheap remediation, but none is needed: the exposure is harmless. A published-but-not-yet-consumed reveal only makes that one future contribution predictable rather than secret, which is no worse than the validator withholding it, and RANDAO tolerates predictable contributions (one honest contributor per mixing period suffices). A validator sufficiently concerned about a specific exposed chain can exit and re-enter, but this is unwarranted in practice.

### Erroneous commitments

Preimage possession is unverifiable at registration: the protocol cannot distinguish a commitment whose chain the validator holds from a typo or the output of a buggy key-derivation.

> Because registration is one-time, a bad commitment cannot be corrected in place: the validator can never propose on the hash-chain path, and the only remedy is to exit and re-enter. This makes an erroneous registration as costly as losing the chain secret, and it does not exist under BLS reveals, where there is nothing to misconfigure beyond the signing key itself. Validator clients SHOULD therefore verify a commitment by walking its full chain before registering, and treat the chain secret with the same care as the signing key.

Reviewer 2: Same here: this `SHOULD` is outside the Specification section, please lowercase to `should` per CONTRIBUTING.md.

The complication is that hash-chain commitments carry no identity and are trivially copyable. A validator can register a value it did not generate by including a copy of another validator's current commitment, via a perfectly valid, freshly signed change. Under a linear XOR accumulator this enables a cancellation attack.
A hash accumulator, `mix = hash(mix + reveal)`, closes this at the root: the hash has no efficiently computable inverse, so re-injecting a copied reveal produces a fresh, unrelated mix rather than undoing anything.

Because the accumulator itself neutralizes copyability, the hash-chain path folds in the raw reveal with no per-validator personalization.

### Why one link per block proposed

The legacy reveal is a signature over the epoch, so two proposals by the same validator in one epoch reveal the same value. Consuming one link per block is simpler (no per-epoch bookkeeping), makes every block contribute fresh entropy, and costs nothing, since chains are cheap to generate at any reasonable length.

> We also note that with consolidations, the probability of a validator proposing multiple times in the same epoch increases, effectively cancelling out their RANDAO contribution.

Reviewer 5: "effectively cancelling out their RANDAO contribution" 

Doesn't this contradict the "hash has no inverse" text above?

Author: Ah this is about the legacy reveal scheme, which uses BLS and xor for accumulations; XORing the same thing together effectively cancels it out

Reviewer 5: Ah makes sense

### Why keep a legacy fallback instead of a registration deadline

There is no urgent need for validators to be fully post-quantum yet, so a hard deadline buys little; unregistered validators keep functioning on the BLS path in the meantime. There will also likely be a long tail of validators who never submit the message because they run custom software. We therefore defer the forced cutover to a fork that must touch these code paths anyway, the post-quantum switchover, which also lets us exercise the new code paths first.

## Backwards Compatibility

This EIP requires a scheduled consensus-layer hard fork. Within the fork, the change is backwards compatible from the validator's perspective: unregistered validators continue proposing exactly as today. Downstream consumers of `randao_mixes` (including the execution layer's `PREVRANDAO`) are unaffected; the mix remains a 32-byte accumulator updated once per block, only the provenance of contributions changes.

## Test Cases

State-transition test vectors to be provided in the consensus-specs test suite.

## Security Considerations

### Biasability is unchanged

> The proposer's only degree of freedom is still withholding: reveal and propose, or withhold and forfeit the block plus its rewards. This is the same one-bit-per-proposer bias RANDAO has today, with the same economic cost. The validator freely choosing its own chain values (whereas a BLS reveal is uniquely determined by the key and epoch, leaving no freedom) does not add grinding power, because the entire chain is fixed before the validator knows its proposal slots or the co-contributions to any future mix.

Reviewer 5: say that this holds assuming collision resistance of the hash function, which is 128 bit classically (for a 256 bit random oracle), but less quantumly.

BLS is not post-quantum safe, so a cryptographically relevant quantum computer (CRQC) can recover a validator's secret key from its public key. Uniqueness still holds, but the attacker can compute the validator's reveals itself, predicting the chain's future randomness, and thus the proposer schedule, far in advance.

Prediction also amplifies bias, beyond merely leaking the schedule. The standard reveal-or-withhold attack requires the attacker to control `k` *consecutive* proposer slots at the tail of an epoch to choose among `2**k` candidate mixes, because it cannot evaluate a candidate without knowing the contributions that land after its own. An attacker that can predict every honest contribution evaluates each candidate outright, so (assuming honest proposers always reveal) any `k` of its slots in the mixing period (the span of slots whose reveals feed the target seed) work; they need not be consecutive or at the end of the epoch.

> Most uses of signatures, such as block proposals and attestations, can move to any secure post-quantum scheme. RANDAO, however, additionally depends on signature *uniqueness*, and most post-quantum schemes do not provide it, allowing one to grind for a favorable RANDAO contribution. The hash-based signature scheme currently planned for the consensus layer (a generalized XMSS, commonly referred to as the lean signature scheme) is grindable in exactly this way, since it includes a *salt* component.

Reviewer 2: ```suggestion
Most uses of signatures, such as block proposals and attestations, can move to any secure post-quantum scheme. RANDAO, however, additionally depends on signature *uniqueness*, and most post-quantum schemes do not provide it, allowing one to grind for a favorable RANDAO contribution. The hash-based signature scheme currently planned for the consensus layer (a generalized eXtended Merkle Signature Scheme (XMSS) ([RFC 8391](https://www.rfc-editor.org/info/rfc8391/)), commonly referred to as the lean signature scheme) is grindable in exactly this way, since it includes a *salt* component.
```

### Why the Protocol Does Not Track Chain Length

Storing only the current commitment and requiring a preimage per proposal makes the chain length a purely private, off-chain parameter. The protocol state cost is one 32-byte entry per validator regardless of chain length, and validators can size chains so that the chain outlasts the validator. The only reason to store the chain length would be to mandate a specific length for every validator, which seems unnecessarily restrictive.

### Why an Activation Delay on Registration, and Why `MIN_SEED_LOOKAHEAD + 2`

The scheme's grinding resistance comes entirely from the commitment predating the validator's knowledge of what it would want to bias. If a validator could register a chain just before proposing and it became active instantly, it could grind: generate many candidate chains, compute the mix each would produce at the upcoming slot, and register the most favourable one. We add a delay, sized so that the registrant cannot even know *whether* it will propose in the activation epoch at the time the registration is included.

> Concretely, proposer duties for epoch `E` are drawn from the mix fixed at the end of epoch `E - MIN_SEED_LOOKAHEAD - 1` (`E - 2` with `MIN_SEED_LOOKAHEAD = 1`). For a registration included in epoch `N`, the seeds for epochs up to `N + 1` are already final, and the seed for `N + 2` (the epoch-`N` mix) is nearly complete for a registrant landing in the last slot of `N`, so a registration activating in any epoch up to `N + 2` is grindable. The seed for `N + 3` fixes only at the end of epoch `N + 1`, entirely after inclusion, making `N + MIN_SEED_LOOKAHEAD + 2 = N + 3` the earliest safe activation epoch; `COMMITMENT_REGISTRATION_DELAY = 3` meets the bound exactly.

Reviewer 2: This part is hard to follow, it defines `MIN_SEED_LOOKAHEAD = 1`. Is this constant an actual constant or could it be changed in the future?
It seems this section explains why the value for `COMMITMENT_REGISTRATION_DELAY` was choosen, but I find it not very clear.

Author: yep this is defined on the CL
